drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb

```diff
@@ -0,0 +1,311 @@
+            log_info("[#{member_row.context},#{posting_row.context}] #{msg}")
+          end
+        end
+      end
+    end
+
+    def build_household_member_records
+      referral_pks_by_id = referral_class.pluck(:identifier, :id).to_h
+      household_member_rows.map do |row|
+        mci_id = row_value(row, field: 'MCI_ID')
+        client_pk = client_pk_by_mci_id(mci_id)
+        unless client_pk
+          log_skipped_row(row, field: 'MCI_ID')
+          next # early return
+        end
+        HmisExternalApis::AcHmis::ReferralHouseholdMember.new(
+          referral_id: referral_pks_by_id.fetch(row_value(row, field: 'REFERRAL_ID')),
+          relationship_to_hoh: relationship_to_hoh(row),
+          mci_id: mci_id,
+          client_id: client_pk,
+        )
+      end
+    end
+
+    def build_posting_records
+      referral_pks_by_id = referral_class.pluck(:identifier, :id).to_h
+      projects_pks_by_id = Hmis::Hud::Project
+        .where(data_source: data_source)
+        .pluck(:project_id, :id)
+        .to_h
+
+      unit_types_by_mper = Hmis::UnitType
+        .joins(:mper_id)
+        .pluck('external_ids.value', :id)
+        .to_h
+
+      posting_rows.map do |row|
+        project_id = row_value(row, field: 'PROGRAM_ID')
+        referral_id = row_value(row, field: 'REFERRAL_ID')
+
+        # find household id for this posting:
+        # 1) find the head-of-household's MCI for this referral
+        hoh_member_row = household_member_rows_by_referral(referral_id).detect do |mr|
+          row_value(mr, field: 'RELATIONSHIP_TO_HOH_ID') == '1'
+        end
+        hoh_mci_id = row_value(hoh_member_row, field: 'MCI_ID')
+        # 2) find the household_id for the project enrollment with that MCI
+        household_id = referral_household_id(hoh_mci_id, project_id)
+        next unless household_id
+
+        HmisExternalApis::AcHmis::ReferralPosting.new(
+          referral_id: referral_pks_by_id.fetch(referral_id),
+          data_source_id: data_source.id,
+          identifier: row_value(row, field: 'POSTING_ID'),
+          status: posting_status(row),
+          project_id: projects_pks_by_id.fetch(project_id),
+          unit_type_id: unit_types_by_mper.fetch(row_value(row, field: 'UNIT_TYPE_ID')),
+          resource_coordinator_notes: row_value(row, field: 'RESOURCE_COORDINATOR_NOTES', required: false),
+          HouseholdID: household_id,
+          status_updated_at: parse_date(row_value(row, field: 'STATUS_UPDATED_AT')),
+          created_at: parse_date(row_value(row, field: 'ASSIGNED_AT')),
+        )
+      end
+    end
+
+    #
+    # Lookups and helpers
+    #
+    def household_member_rows_by_referral(referral_id)
+      @household_member_rows_by_referral ||= household_member_rows.group_by { |row| row_value(row, field: 'REFERRAL_ID') }
+      @household_member_rows_by_referral.fetch(referral_id)
```

Comment:
@ttoomey I'm getting a KeyError here and having a hard time following what's going on. What does it mean here if the referral id isn't found?

Reply:
It seems like this must be a referral id that has no household members in the data set itself. So bad data I guess? I'll adjust it to return an empty set in that case.